arch/eventloop_zephyr/eventloop_zephyr.h

```diff
@@ -0,0 +1,243 @@
+    }
+#define UA_LOG_SOCKET_ERRNO_GAI_WRAP(LOG)                                                \
+    {                                                                                    \
+        const char *errno_str = UA_clean_errno(gai_strerror);                            \
+        LOG;                                                                             \
+        errno = 0;                                                                       \
+    }
+
+/***********************/
+/* General Definitions */
+/***********************/
+
+#define UA_MAXBACKLOG 100
+#define UA_MAXHOSTNAME_LENGTH 256
+#define UA_MAXPORTSTR_LENGTH 6
+
+#ifndef MSG_NOSIGNAL
+#define MSG_NOSIGNAL 0
+#endif
+
+#ifndef MSG_DONTWAIT
+#define MSG_DONTWAIT 0
+#endif
+
+/* Zephyr events are based on sockets / file descriptors. The EventSources can
+ * register their fd in the EventLoop so that they are considered by the
+ * EventLoop dropping into "poll" to wait for events. */
+
+/* TODO: Move the macro-forest from /arch/<arch>/ua_architecture.h */
+
+typedef UA_socket_fd UA_fd;
+#define UA_INVALID_FD UA_INVALID_SOCKET
+
+struct UA_RegisteredFD;
+typedef struct UA_RegisteredFD UA_RegisteredFD;
+
+/* Bitmask to be used for the UA_FDCallback event argument */
+#define UA_FDEVENT_IN 1
+#define UA_FDEVENT_OUT 2
+#define UA_FDEVENT_ERR 4
+
+typedef void (*UA_FDCallback)(UA_EventSource *es, UA_RegisteredFD *rfd, short event);
+
+struct UA_RegisteredFD {
+    UA_DelayedCallback dc; /* Used for async closing. Must be the first member
+                            * because the rfd is freed by the delayed callback
+                            * mechanism. */
+
+    ZIP_ENTRY(UA_RegisteredFD) zipPointers; /* Register FD in the EventSource */
+    UA_fd fd;
+    short listenEvents; /* UA_FDEVENT_IN | UA_FDEVENT_OUT*/
+
+    UA_EventSource *es; /* Backpointer to the EventSource */
+    UA_FDCallback eventSourceCB;
+};
+
+enum ZIP_CMP
+cmpFD(const UA_fd *a, const UA_fd *b);
+typedef ZIP_HEAD(UA_FDTree, UA_RegisteredFD) UA_FDTree;
+ZIP_FUNCTIONS(UA_FDTree, UA_RegisteredFD, zipPointers, UA_fd, fd, cmpFD)
+
+/* All ConnectionManager in the Zephyr EventLoop can be cast to
+ * UA_ConnectionManagerZephyr. They carry a sorted tree of their open
+ * sockets/file-descriptors. */
+typedef struct {
+    UA_ConnectionManager cm;
+
+    /* Statically allocated buffers */
+    UA_ByteString rxBuffer;
+    UA_ByteString txBuffer;
+
+    /* Sorted tree of the FDs */
+    size_t fdsSize;
+    UA_FDTree fds;
+} UA_ZephyrConnectionManager;
+
+typedef struct {
+    UA_EventLoop eventLoop;
+
+    /* Timer */
+    UA_Timer timer;
+
+    /* Linked List of Delayed Callbacks */
+    UA_DelayedCallback *delayedCallbacks;
+
+    /* Flag determining whether the eventloop is currently within the
+     * "run" method */
+    volatile UA_Boolean executing;
+
+    UA_RegisteredFD **fds;
+    size_t fdsSize;
+
+    /* Self-pipe to cancel blocking wait */
```

Comment:
If you only need the self-pipe to abort an (endless) call to poll/select, Zephyr has so-called event-fd.
https://docs.zephyrproject.org/latest/doxygen/html/posix_2sys_2eventfd_8h.html

They event-fd can also be listened on with poll/select.
That will have less overhead and is generally cleaner than sending bits over a localhost udp connection.

Reply:
Nice hint. Will check this out.

Reply:
done